core/tx_pool.go

```diff
@@ -916,6 +941,13 @@ func (pool *TxPool) addTxs(txs []*types.Transaction, local, sync bool) []error {
 		return errs
 	}
 
+	// Track private transactions, so they don't get leaked to the public mempool
+	if private {
+		for _, tx := range news {
+			pool.privateTxs.Add(tx.Hash())
+		}
+	}
```

Comment:
I think there's a possible memleak here. If the same tx is added twice (either it is present twice in the `news`, or it's added a second time. Then it will be added to the internal list twice. 
So list is `[ txhash, txhash, ...]`. timestamps are `{ txhash: last_timestamp, ..}`. 
After pruning, the first txhash will be removed, and the timestamp is removed. 

Next time pruning happens, the timestamp lookup will return `0`, meaning that no more pruning will ever occur. If there are other txs that slipped in in-between, the `0`-timestamp-element will sit "unsorted", and eventually become the first element, causing pruning to halt. 


This can be made self-healing, if instead of storing timestamp of "when to delete", you change it to "when was it added". Because then, if there's a mismatch and the timestamp is missing, the value `0` will cause it to be pruned.

Reply:
:+1: